go/plugins/googleai/googleai.go

```diff
@@ -16,17 +16,154 @@ package googleai
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"path"
+	"slices"
+	"sync"
 
 	"github.com/firebase/genkit/go/ai"
 	"github.com/firebase/genkit/go/plugins/internal/uri"
 	"github.com/google/generative-ai-go/genai"
+	"golang.org/x/exp/maps"
 	"google.golang.org/api/iterator"
 	"google.golang.org/api/option"
 )
 
-func newClient(ctx context.Context, apiKey string) (*genai.Client, error) {
-	return genai.NewClient(ctx, option.WithAPIKey(apiKey))
+var state struct {
+	mu        sync.Mutex
+	client    *genai.Client
+	models    map[string]ai.Generator
+	embedders map[string]ai.Embedder
+}
+
+// Config configures the plugin.
+type Config struct {
+	// API key. Required.
+	APIKey string
+	// Generative models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Models []string
+	// Embedding models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Embedders []string
+}
+
+func Init(ctx context.Context, cfg Config) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("googleai.Init: %w", err)
+		}
+	}()
+
+	if err := initClient(ctx, cfg.APIKey); err != nil {
+		return err
+	}
+
+	needModels := len(cfg.Models) == 0
+	needEmbedders := len(cfg.Embedders) == 0
+	if needModels || needEmbedders {
+		iter := state.client.ListModels(ctx)
+		for {
+			mi, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			// Model names are of the form "model/name".
+			name := path.Base(mi.Name)
```

Comment:
This isn't really a path, so I think it would be easier to understand as

    name, _, _ := strings.Cut(mi.Name, "/")

Reply:
(a) It sort of is. [GCP resource names are slash-separated paths](https://cloud.google.com/apis/design/resource_names).

(b) We want the last component, not the first, so the code wouldn't be as nice.

I think this is fine.